staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package responsewriter
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+)
+
+// Decorator composes an inner http.ResponseWriter object, the decorator
+// object should return the inner (decorated) http.ResponseWriter
+// object via the Unwrap function.
+type Decorator interface {
+	http.ResponseWriter
+
+	// Unwrap returns the inner http.ResponseWriter object associated with the decorator
+	Unwrap() http.ResponseWriter
+}
+
+// Wrap accepts a Decorator object and wraps with an "outer" http.ResponseWriter
+// object that implements http.CloseNotifier, http.Flusher and http.Hijacker accordingly
+//
+// We use three terms here:
+//  - "outer": this is the ResponseWriter oject returned by this function. The outer object
+//    must implement the same exact interfaces as does the inner object. Our scope is limited
+//    to http.CloseNotifier, http.Flusher and http.Hijacker. This function does not take into
+//    account any other interfaces a ResponseWriter object may implement.
```

Comment:
Since this function might return the middle object, defining "outer" to be what this function returns is a little problematic because it means "outer" = "middle" in that case.

We may not need a term that means "what this func returns" --- we can just say that when that is what we mean.

What about saying something like the following?

> // Wrap accepts an `http.ResponseWriter` and potentially wraps it with another object that
> // implements `http.CloseNotifier` and `http.Flusher` and possibly `http.Hijacker`
> // by delegating to an inner `http.ResponseWriter`, so that the returned `http.ResponseWriter`
> // implements the same subset of those interfaces as the inner `http.ResponseWriter`.
> //
> // We use three terms here:
> // - "outer": this is a ResponseWriter object introduced by this function, adding delegation
> //     to the inner ResponseWriter.
> // - "middle": this is the ResponseWriter object from the client ...

Reply:
done